File: python/level2/sort_42747.py
# https://programmers.co.kr/learn/courses/30/lessons/42747

# sort를 하면 citations 길이 - idx가 h번 이상 인용된 논문의 갯수라는 점이 핵심
def solution(citations):
    citations.sort()
    l = len(citations)
    for i in range(l):
        if citations[i] >= l - i:
            # 논문이 인용된 횟수(h번 이상) >= 인용된 논문의 개수(h개 == h번)
            return l - i
    return 0

citations = [3, 0, 6, 1, 5]
print(solution(citations))

File: java/src/solution/level2/sort_42747.java
// https://programmers.co.kr/learn/courses/30/lessons/42747

package solution.level2;
import java.util.*;

public class sort_42747 {
    public static int solution(int[] citations) {
        int answer = 0; // {0, 0, 0, 0, 0}일떄는 이 값이 출력됨
        Arrays.sort(citations); // citations를 정렬했기 때문에 아래 for문의 logic으로 h-index 산출가능
        // 순서대로 정렬했기 때문에 h = citations.length - i 가 되야함

        for (int i = 0; i < citations.length; i++) {
            int h = citations.length - i; // h구하는게 핵심

            if (citations[i] >= h) {
                answer = h;
                break;
            }
        }
        return answer;
    }
    public static void main(String[] args){
        int[] citations = {3, 0, 6, 1, 5};
        Integer answer = solution(citations);
        System.out.println(answer);
    }
}
